deviceModeInit.js

```diff
@@ -117,16 +115,143 @@ var rudderTracking = (function () {
+      const productUrlRegex = new RegExp(
+        `^(?!\/\/cdn)[-\.:\/,a-z,A-Z,0-9]*\/products\/`
+      );
+      const allAnchorTagsWithProducts = Array.prototype.slice
+        .call(allAnchorTags)
+        .filter((anchorTag) => {
+          return productUrlRegex.test(anchorTag.href);
+        });
+      let allProductsOnPageWithImg = allAnchorTagsWithProducts.filter(
+        (anchorTag) => {
+          anchorTag.querySelector("img");
+        }
+      );
+      if (allProductsOnPageWithImg.length === 0) {
+        allProductsOnPageWithImg = allAnchorTagsWithProducts.filter(
+          (anchorTag) => {
+            const parentNode = anchorTag.parentNode.parentNode.parentNode;
+            return (
+              parentNode.nextElementSibling?.querySelector("img") ||
+              parentNode.previousElementSibling?.querySelector("img")
+            );
+          }
+        );
+      }
+      return allProductsOnPageWithImg;
+    }
+    function callProductListViewedEvent(productsOnThePage, productConsidered) {
+      const elementsToSend = [];
+      const productsToSend = [];
+      productsOnThePage.forEach((element) => {
+        const isProductVisible = productIsVisible(window, element);
+        const isProductAlreadyConsidered = productConsidered.includes(
+          element.href
+        );
+        if (isProductVisible && !isProductAlreadyConsidered) {
+          elementsToSend.push(element);
+          productConsidered.push(element.href);
+        }
+      });
+      if (elementsToSend.length > 0) {
+        elementsToSend.forEach(async (product) => {
+          const handle = product.href.match(
+            /(\/products\/)((\w|-)*)(\?|\$?)/
+          )[2];
+          await rs$
+            .get(`/products/${handle}.json`, undefined, undefined, "JSON")
+            .then((data) => {
+              const products = [];
+              const rudderstackProduct = propertyMapping(
+                data.product,
+                productMapping
+              ); // here as well 
+              rudderstackProduct.currency = pageCurrency;
+              rudderstackProduct.sku = String(
+                rudderstackProduct.variant?.sku ||
+                rudderstackProduct.product_id
+              );
+              rudderstackProduct.price = rudderstackProduct.variant?.price;
+              products.push(rudderstackProduct);
+              productsToSend.push(rudderstackProduct);
+            })
+            .catch((error) => {
+              console.debug("Rudderstack unable to fetch", handle, error);
+            });
+        });
+        window.setTimeout(() => {
+          if (productsToSend.length > 0) {
+            rudderanalytics.track("Product List Viewed", {
+              products: productsToSend,
+            });
+          }
+        }, 2500);
+      }
+    }
+    function productListViews() {
+      const productsOnThePage = getAllProductsOnPage();
+      const productConsidered = [];
```

Comment:
Shouldn't this be a global variable?

Reply:
@ItsSudip would be more suitable for this